src/schema.rs

```diff
@@ -321,30 +342,51 @@ impl RecordField {
 pub struct UnionSchema {
     schemas: Vec<Schema>,
     // Used to ensure uniqueness of schema inputs, and provide constant time finding of the
-    // schema index given a value.
-    // **NOTE** that this approach does not work for named types, and will have to be modified
-    // to support that. A simple solution is to also keep a mapping of the names used.
+    // schema index given a name or value.
+    name_index: HashMap<String, usize>,
     variant_index: HashMap<SchemaKind, usize>,
 }
 
 impl UnionSchema {
     pub(crate) fn new(schemas: Vec<Schema>) -> Result<Self, Error> {
         let mut vindex = HashMap::new();
+        let mut nindex = HashMap::new();
+
         for (i, schema) in schemas.iter().enumerate() {
-            if let Schema::Union(_) = schema {
-                Err(ParseSchemaError::new(
-                    "Unions may not directly contain a union",
-                ))?;
-            }
-            let kind = SchemaKind::from(schema);
-            if vindex.insert(kind, i).is_some() {
-                Err(ParseSchemaError::new(
-                    "Unions cannot contain duplicate types",
-                ))?;
+            match schema {
+                Schema::Union(_) => {
+                    Err(ParseSchemaError::new(
+                        "Unions may not directly contain a union",
+                    ))?;
+                }
+
+                // named types
+                Schema::Record { name, .. } => {
```

Comment:
I believe that the complete list of name types are Record, Fixed and Enum (see #95  for a quote taken from the spec). Would you have time to add support also for the other two missing types?

Reply:
I don't mind adding it. I wanted to make the least amount of changes to the code base to keep the PR small but it makes sense to add these in now anyways.

Reply:
I very much appreciate that! Long PRs are a ton of work and it is hard to get them merged. Small ones like this one are much better! :)